core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -0,0 +1,500 @@
+        NodeWithSymbols second = appendHelperSymbolsForSource(1, node, partitionSizeSymbols, rowNumberSymbolsBuilder, metadata, context);
+        JoinedNodes joined = join(first, second, context);
+
+        for (int i = 2; i < node.getSources().size(); i++) {
+            NodeWithSymbols joinedWithSymbols = appendHelperSymbolsForJoinedNodes(joined, partitionSizeSymbols, combinedRowNumberSymbols, context);
+            NodeWithSymbols next = appendHelperSymbolsForSource(i, node, partitionSizeSymbols, rowNumberSymbolsBuilder, metadata, context);
+            joined = join(joinedWithSymbols, next, context);
+        }
+
+        Map<Symbol, Symbol> rowNumberSymbols = rowNumberSymbolsBuilder.buildOrThrow();
+
+        NodeWithSymbols joinedWithSymbols = appendHelperSymbolsForJoinedNodes(joined, partitionSizeSymbols, combinedRowNumberSymbols, context);
+        // The max row number symbol from all joined partitions.
+        Symbol finalRowNumberSymbol = joinedWithSymbols.getRowNumber();
+
+        NodeWithMarkers marked = appendMarkerSymbols(joinedWithSymbols.getNode(), ImmutableSet.copyOf(rowNumberSymbols.values()), finalRowNumberSymbol, context);
+
+        // Remap the symbol mapping: replace the row number symbol with the corresponding marker symbol.
+        // In the new map, every source symbol is associated with the corresponding marker symbol.
+        // Null value of the marker indicates that the source value should be ignored by the table function.
+        ImmutableMap<Symbol, Symbol> markerSymbols = rowNumberSymbols.entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> marked.getSymbolToMarker().get(entry.getValue())));
+
+        // Keep the final row number symbol, and use it for ordering the combined sources.
+        // It runs along each partition in the cartesian product, numbering the partition's rows according to the expected ordering / orderings.
+        // note: ordering is necessary even if all the source tables are not ordered. Thanks to the ordering, the original rows
+        // of each input table come before the "filler" rows.
+        Optional<OrderingScheme> finalOrderBy = Optional.of(new OrderingScheme(ImmutableList.of(finalRowNumberSymbol), ImmutableMap.of(finalRowNumberSymbol, ASC_NULLS_LAST)));
+
+        // Partition the combined sources by their combined partitioning lists.
+        List<Symbol> partitionBy = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getSpecification)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(DataOrganizationSpecification::getPartitionBy)
+                .flatMap(Collection::stream)
+                .collect(toImmutableList());
+
+        PlanNode pruned = pruneHelperSymbols(marked.getNode(), partitionSizeSymbols.build(), rowNumberSymbols.values(), combinedRowNumberSymbols.build(), finalRowNumberSymbol, context);
```

Comment:
Why do we need this? Any outputs of `marked.getNode()` that are not referenced by the resulting `TableFunctionProcessingNode` should be pruned independently by the pruning optimizer rules.